File: frontend/src/services/api/novel-promotion/projects.ts
import type { NPProject, NPProjectUpdate } from '../../../types/novel-promotion'
import { apiClient } from '../client'

export async function getNPProject(projectId: string) {
  const response = await apiClient.get(`/novel-promotion/${projectId}`)
  return response.data.data.np_project as NPProject
}

export async function updateNPProject(projectId: string, payload: NPProjectUpdate) {
  const response = await apiClient.patch(`/novel-promotion/${projectId}`, payload)
  return response.data.data.np_project as NPProject
}

File: backend/app/api/routes/np/projects.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps.auth import get_current_user
from app.core.db import get_db
from app.db.models.user import User
from app.schemas.novel_promotion import (
    NovelPromotionProjectRead,
    NovelPromotionProjectUpdate,
)
from app.services.novel_promotion import NovelPromotionProjectService

router = APIRouter()


@router.get("/{project_id}")
def get_np_project(
    project_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = NovelPromotionProjectService(db)
    np = service.get_or_create(current_user.id, project_id)
    return {
        "success": True,
        "data": {"np_project": NovelPromotionProjectRead.model_validate(np).model_dump()},
    }


@router.patch("/{project_id}")
def update_np_project(
    project_id: str,
    payload: NovelPromotionProjectUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = NovelPromotionProjectService(db)
    np = service.update(current_user.id, project_id, payload)
    return {
        "success": True,
        "data": {"np_project": NovelPromotionProjectRead.model_validate(np).model_dump()},
    }
